### ISSUE TITLE
Introducing Skov

### ISSUE THREAD

**Issue by nicolas-p** [nicolas-p/skov#1]
@mrjbq7, @erg, @bjourne,

I am starting a new project based on Factor. Skov is a visual (or graphical) programming system where programs are represented by graphs. It can be thought of as a visual Lisp.

I had been thinking of a visual Lisp for some time because Lisp programs are trees and I thought that trees could be displayed better in a graphical form rather than as text. At the time, I hadn't understood all the advantages of this so I didn't pursue the idea.

In early August, I realised that Lisp, Factor and Rebol, three very powerful languages, are all solving problems present in the other two while introducing new problems. They are sort of circling around an ideal without being able to get there. Lisp explicitely shows the structure of the program, at the cost of lots of parentheses. Factor and Rebol don't need the parentheses but the structure of the code is lost. The structure has to be recreated by the compiler and by the programmer's brain. Each word has to take a fixed number of arguments and the compiler and the programmer have to know the number of arguments taken by each one. Factor is written in the order of execution, which allows it to pass data between words without the need for named arguments. The drawback of this is the profusion of maddening stack sufflers. Lisp and Rebol don't need stack shufflers but they require the programmer to give names to everything. This prevents fine-grained refactoring, which is also limited in Factor by the need to write stack effects. What prevents Lisp, Factor and Rebol from reaching the ideal is text-based input.

I realised that it is not possible to have a text-based language where code is easy to write, easy to read, program structure is explicit, refactoring is unlimited, results can be easily used in several places, data can be passed around easily, and where the programmer doesn't have to remember anything other than function names.

Consider what a visual system allows us to get rid of: _let_ expressions, parentheses, stack shufflers, positional arguments, all syntax.

I researched the field of visual languages, which was mostly unknown to me. In most systems, the programmer has to move blocks or nodes around to draw his program, almost spending more time doing interior decoration than actual programming. In Skov, the user doesn't have any control over the appearance of the program. The graph is automatically layed out.

On the question of productivity, I would argue that programming in Skov may actually be faster than in text-based languages. You enter all the words you will be needing with your keyboard, in the order in which they come to mind, not in an order imposed by the system. Then you take your mouse and simply draw all the connections between the words.

Factor was the best possible base for Skov because it is a powerful language, it has a lot of vocabularies, words have good names, it has image-based persistance and an OpenGL user interface, it works the same on the three common platforms and it is entirely contained in one directory.

The Skov project currently contains about 800 lines of Factor code, which is amazingly little.

At the moment, it is possible to create vocabularies and words, to use words in other words, to execute words and see the results. It is possible to use words that take quotations like _map_, _filter_ or _if_. You can save the image so that your work will still be there the next time you open Skov.

In the future, I imagine there could be a _system_ image that the user wouldn't modify and several user-created images that would contain the Skov code. These images could be sent to other people to share code. I would also like Skov to be homoiconic, not in the sense of code and data having the same appearance, but in the sense of them being processed by similar tools.

Here is what the interface looks like:
![capture d ecran 2015-10-22 a 20 05 05](https://cloud.githubusercontent.com/assets/11431580/10673934/aed51b62-78f8-11e5-8c03-96b19671f3e1.png)

I would be happy to read your comments on this project.

**Comment by mrjbq7** [comment]
This is cool!

I'm not sure how to draw "links", or move the colored boxes around.  Tried adding a few words and inputs and outputs but not sure how to make them connect.

Also, have you seen this demo before?

https://www.youtube.com/watch?v=R3MNcA2dpts

**Comment by erg** [comment]
I've always thought radial buttons would be good as input for Factor, and someone implemented a concept last month on hackernews:

https://github.com/pschanely/wildflower-touch

**Comment by mrjbq7** [comment]
Also, I see copyright but no license, what license are you making this available with?

Would you like some contributions?  I will have some time to browse the code a bit later and give you some pull requests you can merge if you want (or ignore if you don't like!).

**Comment by nicolas-p** [comment]
To draw links, click a connector and hold the button down as you move the cursor to another connector.

**Comment by nicolas-p** [comment]
John, I have just watched the video on Brief that you suggested. I have tried to reproduce this example:

![capture d ecran 2015-10-24 a 16 39 12](https://cloud.githubusercontent.com/assets/11431580/10710982/74532a98-7a6e-11e5-86f0-373516656dbc.png)

This is my version:

![capture d ecran 2015-10-24 a 18 08 13](https://cloud.githubusercontent.com/assets/11431580/10711399/536dea6e-7a7a-11e5-8e60-265da12b9fc0.png)

I think it shows an advantage of this system: the number 9 can be used twice without the need for the _bi_ combinator and its quotations.

To answer your other question, contributions are of course welcome. There is one point in particular that might be worth looking at. In the code, I haven't used Factor's _models_ vocabulary. Instead, I added a `modell` (with two Ls, yes, this is ugly) attribute to my gadgets because I thought this was simpler. I'm still wondering if this was the best thing to do or if I should have used the _models_ vocabulary.

**Comment by nicolas-p** [comment]
It is now possible to use `each`:

![capture d ecran 2015-11-06 a 22 00 54](https://cloud.githubusercontent.com/assets/11431580/11008643/2f0a07e0-84d2-11e5-989c-01e874f2b32f.png)

This outputs:

```
30
40
50
```

And more generally, to chain words that are only useful for their side effects:

![capture d ecran 2015-11-06 a 22 07 11](https://cloud.githubusercontent.com/assets/11431580/11008738/e4458b52-84d2-11e5-990b-3ae4b3afe181.png)

**Comment by mrjbq7** [comment]
This project is really cool. I've been busy but will look at it hopefully soon-ish.

**Comment by nicolas-p** [comment]
Thank you John.

I've implemented most of what I had in mind when I had the initial idea three months ago. Now I need to think about what to do next.

Some things I would like to improve:
- when a new word is created it should be selected automatically, this is harder to do than it seems
- `while` and `until` don't work and I haven't yet found a way to make them work
- it is not currently possible to define classes (`TUPLE`, `MIXIN`, `UNION`, etc) and methods

I think that the biggest issue I will have to solve in the future is how to move programs around (how to send them to other people, and how to use them in a newer version of Skov). Because currently, the programs are saved in the image and stuck in it forever.

**Comment by mrjbq7** [comment]
It should be pretty easy to use `see` or some of similar machinery to "export" the defined words to a file for sharing.  We have plans after 0.98 release with the new parser to make two-way editing easier (for example when changing or defining words in the image, propagating that change to the appropriate files).

It seems going for a subset of Factor would be appropriate here, as for `while` and `until`, those might need special handling in the UI, perhaps the issue has to do with the "row polymorphic" stack effects of the quotations they take, `( ..a pred: ( ..a -- ..b ? ) body: ( ..b -- ..a ) -- ..b )`?

**Comment by nicolas-p** [comment]
Your last answer made me think about some alternative architectures for this project:

![architectures](https://cloud.githubusercontent.com/assets/11431580/11194746/22c2cd3a-8cae-11e5-99e0-27ca38c597a5.png)

The last option gets rid of Factor code completely and needs a modified virtual machine and compiler. All Factor words would be converted to the graph representation.

This is just to organise my thoughts, it is not a roadmap.

**Comment by nicolas-p** [comment]
It is now possible to define recursive functions. This is the classical example, the factorial function:

![capture d ecran 2015-12-06 a 20 32 16](https://cloud.githubusercontent.com/assets/11431580/11640224/bfff1c50-9d31-11e5-93c6-96471bf7c7c4.png)

Looks frightening at first sight, but if you look at it from bottom to top, it reads like Lisp code.

**Comment by mrjbq7** [comment]
Wow, thats really cool.

**Comment by mrjbq7** [comment]
I need to look at this more soon, been way too busy last few weeks.

**Comment by nicolas-p** [comment]
At last, it is possible to use `while` and `until`, something I had been wanting for some time.

![capture d ecran 2016-01-06 a 22 23 30](https://cloud.githubusercontent.com/assets/11431580/12155254/92df3200-b4c4-11e5-86a0-102db9846f20.png)

This example prints `"eeeeeeeee"`

**Comment by mrjbq7** [comment]
Is there an implicit `dup`?

Isn't that equivalent to:

``` factor
"e" [ dup length 10 = ] [ "e" append ] until
```

**Comment by nicolas-p** [comment]
Yes, there is an implicit `dup`.

By the way, if I remember correctly, there is no need for `dup` when using `while` in the Joy language. When I learned Factor I was surprised by this need for `dup`, but I guess there is a good technical reason for this.

**Comment by mrjbq7** [comment]
Wouldn't it be confusing if the while/until predicate took multiple values?

Or do you implicitly dup the number of inputs the predicate takes?

**Comment by mrjbq7** [comment]
You can use the `preserving` word for that, or build a "smart" combinator:

``` factor
"e" [ [ length 10 = ] preserving ] [ "e" append ] until
```

something like this maybe for two arguments:

``` factor
"e" 5 [
    [ [ length 10 = ] dip zero? or ] preserving
] [
    [ "e" append ] dip 1 -
] until
```

**Comment by nicolas-p** [comment]
Can you give me a good example where `until` needs to work on two values instead of one?

**Comment by mrjbq7** [comment]
Well, I mean if you need multiple values you can always pack them all into a tuple and then act on a single object.  It's your project, but I think doing automatically assuming and preserving the top of the stack as the condition object the predicate acts on is a little strange.  If you don't do anything do you have issues with growing the stack?

**Comment by mrjbq7** [comment]
Basically, the stack effect of `until` is currently:

``` factor
: until ( ..a pred: ( ..a -- ..b ? ) body: ( ..b -- ..a ) -- ..b )
```

And you want it to be?

``` factor
: until ( ..a obj pred: ( ..a obj -- ..b ? ) body: ( ..b obj -- ..a obj ) -- ..b )
```

**Comment by nicolas-p** [comment]
I've just tried to create a word that uses `until` with a predicate or a body that takes two values. It just doesn't compile because my special `until` has the following stack-effect:

```
( initial pred: ( a -- ? ) body: ( b -- a ) -- final )
```

So there can't be any bad behaviour with the stack.

For the moment I'm fine with having `until` and `while` working on a single value. But if I decide to allow a variable number of inputs in the future, I will remember to be careful about preserving values on the stack.

**Comment by mrjbq7** [comment]
Why wouldn't that compile?

``` factor
: special-until ( initial pred: ( a -- ? ) body: ( b -- a ) -- final )
    [ [ preserving ] curry ] dip until ; inline
```

Using it:

```
IN: scratchpad 10 [ 0 < ] [ dup . 1 - ] special-until
10
9
8
7
6
5
4
3
2
1
0

--- Data stack:
-1
```

**Comment by nicolas-p** [comment]
You misunderstood me but it's my fault. What doesn't compile is this:

![capture d ecran 2016-01-08 a 19 07 17](https://cloud.githubusercontent.com/assets/11431580/12205498/5bbba0b0-b63b-11e5-9d86-02e8fafa0db0.png)

(where the `until` node is actually executed as `special-until`)
The fact that it doesn't compile is an expected behaviour.

Your `special-until` looks better than mine so I will probably use it instead. But I found I had to inline it otherwise it says:

```
Cannot apply “call-n” to an input parameter of a non-inline word
```

This may be because I'm still on the last 64-bit Factor for Mac available on the website (mid-August). There has been a lot of changes since.

**Comment by mrjbq7** [comment]
(Yeah I meant to inline it, and updated my post above).

**Comment by nicolas-p** [comment]
I made a binary release available here on GitHub. There's a Mac and a Windows version (which doesn't look as good because of the transparency bug).

This is what I'm planning for the near future:
- adding the possibility of defining tuple classes, with automatic generation of constructors, accessors, mutators and "destructors" (functions that take the object apart and return all the attributes)
- exploring the possibility of having user code in a separate image, even several separate images.

**Comment by nicolas-p** [comment]
This is how hyperbolic sine and cosine could be defined:

![cosh sinh](https://cloud.githubusercontent.com/assets/11431580/13205299/9db6a75e-d8e5-11e5-980c-85a26f8ac3d1.png)

**Comment by mrjbq7** [comment]
You might think of writing all your code back to a file so you can re-use your work from one binary to the next.

Something like:

``` factor
"/path/to/stuff.factor" utf8 [
    "scratchpad" vocab-words
    [ [ name>> "DEFER: %s\n" printf ] each nl ]
    [ [ see nl ] each ] bi
] with-file-writer
```

**Comment by mrjbq7** [comment]
Then to load it:

``` factor
"/path/to/stuff.factor" parse-file drop
```

**Comment by nicolas-p** [comment]
I've just tried your code. As I thought, it works for the "Factor words", but it doesn't save the "Skov words". Factor code can always be re-generated from Skov's internal graph representation. What is currently missing is a way of transferring this graph representation from one release to another. 

What I had in mind since almost the beginning was to use several images:
- _skov.image_ would contain all the Factor code written by Factor developers and myself
- _user.image_ would contain the Skov code "written" by the user
- the user would be able to create more images, for instance to share a vocabulary with someone else
- if the user receives an image from someone, he would just place it in the Skov directory and the Skov code it contains would be loaded at startup

So binary images would replace text files in this system, as a way to store and share code.

Is there any way to do this? Or is Factor only able to work on a single image? I've actually no idea of how Factor works at this low level.

**Comment by mrjbq7** [comment]
Hmm, I haven't looked deeply enough to know what "Skov words" are, but maybe later I could help with how to reconstruct them.

At the moment, Factor is saved and loaded from a single image file, but I would guess that with some work we could support multiple images... but not sure how much work.

**Comment by mrjbq7** [comment]
If you can extract the Skov code into a set of Factor instructions that recreate it, then save those instructions to disk, then you have everything you need.

**Comment by nicolas-p** [comment]
Yes, this is how I worked at the beginning of the project when the UI could only display code but not edit it. I had to write the graph as Factor instructions in a text file.

If it turns out it's too difficult to support multiple images, this could be a backup solution.

To give you a very short explanation of how the graph is represented internally, it's quite simple: I have `vocab` objects, `word` objects and `connector` objects (either inputs or outputs). Connectors are inside words and they have a `link` slot that points to another connector (in another word).

**Comment by nicolas-p** [comment]
Until now, to program in Skov, you had to either:
- add two words, then connect them together, add another word, connect it, add another word, connect it, and so on;
- or add all the words one by one, and then draw all the connections one by one.

Entering all the words is not a problem but I found the connecting phase rather annoying. 

Since the last commit, by placing the cursor over a connector and pressing W, a new word is added with a connection already made from this connector. This makes programming in Skov much faster and nicer.

**Comment by mrjbq7** [comment]
You've been making great progress!

I really am enjoying the programming paradigm you're experimenting with.  I'd love to make deploying easier for you... if you have upstream changes that you'd benefit from, please ask!!

**Comment by nicolas-p** [comment]
Well, currently, I would benefit from fixing [this](https://github.com/factor/factor/issues/1529) and [this](https://github.com/factor/factor/issues/1397).

**Comment by nicolas-p** [comment]
I have just thought of an improvement that can be made in Factor's help browser. Instead of displaying inputs and outputs together, like this:

![capture d ecran 2016-03-18 a 23 19 12](https://cloud.githubusercontent.com/assets/11431580/13893689/02e90bf0-ed61-11e5-8632-00e42ca85e9a.png)

It would be better to show this:

![inputs-outputs](https://cloud.githubusercontent.com/assets/11431580/13893703/13c5772e-ed61-11e5-9544-540aba486241.png)

I wanted to do it but it is harder than I thought so if you think it's a good idea, you can implement it.

**Comment by mrjbq7** [comment]
Looks nice, what if an input and output are the same, duplicate it (once in each section)?

**Comment by nicolas-p** [comment]
Ah, I hadn't thought of that. I see it happens for destructive operations like `remove!` and `append!`.

For Skov, I am not sure these destructive operations would be useful, so I really want to separate inputs and outputs.

For Factor... I don't know, it may not be such a good idea after all. Or inputs and outputs could be displayed separately for the majority of words, but for destructive operations they would be displayed together.

**Comment by mrjbq7** [comment]
I'm not opposed to the idea at all.  Your example looks a lot nicer.  Having it be duplicated isn't a terrible thing...?

**Comment by nicolas-p** [comment]
OK, we can duplicate it. 
If the stack effect is `( elt seq -- seq )`, it is logical to show:

**Inputs**
elt - an object
seq - a resizable mutable sequence

**Outputs**
seq - a resizable mutable sequence

**Comment by nicolas-p** [comment]
I abandoned the idea of using images to share code. Now when you press Ctrl+S, the code is saved as text files in the _work_ directory, and when you press Ctrl+L, the files are loaded back into the image. So you can now share code with other people, carry it between releases, post it on the Internet, use version control... all the things you would expect from a normal programming language.

**Comment by nicolas-p** [comment]
Skov now has full support for high resolution displays and it's GORGEOUS!

<img width="489" alt="capture d ecran 2016-04-30 a 23 13 33" src="https://cloud.githubusercontent.com/assets/11431580/14938618/b1a6a9ce-0f29-11e6-883c-07de171d45c4.png">

**Comment by mrjbq7** [comment]
!!!

**Comment by nicolas-p** [comment]
I've started working on the auto-completion feature. It has two purposes:
- you can type the beginning of a word, press TAB, and select the word you want in the list that appears
- if you have typed a name that matches several words, defined in different vocabularies (like "replace"), the system will ask you to choose the one you want

<img width="676" alt="capture d ecran 2016-07-28 a 22 12 29" src="https://cloud.githubusercontent.com/assets/11431580/17227806/6a16d746-5510-11e6-917e-774714be2f25.png">

**Comment by nicolas-p** [comment]
New milestone: Factor code can now be automatically converted to Skov code (quotations are not yet supported).

So we can display the Skov version of some words that have been in the Factor image forever:

``` factor
\ 1quotation word-definition-from-factor <model> <graph-gadget> gadget.
```

<img width="163" alt="capture d ecran 2016-08-01 a 17 10 18" src="https://cloud.githubusercontent.com/assets/11431580/17300785/21840cea-5814-11e6-9672-315999804d0d.png">

``` factor
\ 1array word-definition-from-factor <model> <graph-gadget> gadget.
```

<img width="115" alt="capture d ecran 2016-08-01 a 17 08 27" src="https://cloud.githubusercontent.com/assets/11431580/17300824/43e3f9f8-5814-11e6-8d6a-276951b98045.png">

``` factor
\ join word-definition-from-factor <model> <graph-gadget> gadget.
```

<img width="194" alt="capture d ecran 2016-08-01 a 17 08 40" src="https://cloud.githubusercontent.com/assets/11431580/17300851/68ececd2-5814-11e6-814e-76fe76a472b3.png">

``` factor
\ glue word-definition-from-factor <model> <graph-gadget> gadget.
```

<img width="237" alt="capture d ecran 2016-08-01 a 17 08 53" src="https://cloud.githubusercontent.com/assets/11431580/17300883/a272768e-5814-11e6-9237-3ca5446837e3.png">

With this, it will be possible to give the illusion that Skov was created from scratch as a visual language, and hide the fact that it is based on Factor.

It will be a 100% visual system where all the code can be browsed and modified, including the UI code, the compiler and all libraries!

**Comment by mrjbq7** [comment]
Wow!!!

**Comment by nicolas-p** [comment]
I've just implemented quotations as well so now we can convert more words:

``` factor
\ clump word-definition-from-factor <model> <graph-gadget> gadget.
```

<img width="227" alt="capture d ecran 2016-08-01 a 22 06 43" src="https://cloud.githubusercontent.com/assets/11431580/17307309/b7730ca4-5834-11e6-8a51-91f90dff388b.png">

``` factor
\ all-eq? word-definition-from-factor <model> <graph-gadget> gadget.
```

<img width="151" alt="capture d ecran 2016-08-01 a 22 07 01" src="https://cloud.githubusercontent.com/assets/11431580/17307301/a5844b7a-5834-11e6-8fd9-54ad32e56da4.png">

I've seen strange results for more complicated words. There are probably a lot of particular cases that need special care (like the `bi` combinator that becomes useless). I'm expecting this automatic conversion to remain partially buggy for a long time. But it's already a huge progress in a single day.

**Comment by nicolas-p** [comment]
I've been struggling with the graph rendering algorithm for the last few months (actually I've been struggling with it since I started the project) but I think I'm getting close now. A graph as complex as this one is now rendered perfectly:

<img width="341" alt="capture d ecran 2016-10-09 a 19 59 32" src="https://cloud.githubusercontent.com/assets/11431580/19222578/08dd0664-8e5c-11e6-8e7a-9b00dbf1ef5b.png">

It's really quite beautiful.

**Comment by mrjbq7** [comment]
That's great! I need to try out your latest version!

**Comment by bjourne** [comment]
Looks cool. Rendering graphs to good looking images is really hard.

**Comment by erg** [comment]
Is it a directed graph? You could experiment with color gradients or have arrows or something. (Or maybe it's always top->bottom?)  Very cool!

**Comment by nicolas-p** [comment]
Arrows were present in my original mock-ups last year but they quickly disappeared. Just imagine that data is subject to gravity like us so it naturally flows from top to bottom. 

As you can see on this page, in terms of aesthetics, there hasn't been any change for many months because I really like it this way and I'm not planning to change anything. The only problem I still have is placing all the nodes in the right places and this is super hard. I feel like I'm really really close now but there are still some bugs.

I would like to work on new features but I really want to finish the graph-layout algorithm first. It just has been way more difficult than I expected.

**Comment by nicolas-p** [comment]
I have greatly improved the [Skov website](http://skov.software/en/). You now have:
- download links
- a tutorial that assumes no previous knowledge of programming but introduces higher order functions, recursion and objects
- an explanation of the fundamental advantages of this type of visual programming over traditional textual languages
- a list of questions and answers

Feedback would be greatly appreciated.

I think websites of software projects generally do a very bad job of introducing their projects. What the visitor wants to know is: What is it? What does it look like? Where can I download it? How do I use it?

I will make a new release available soon (hopefully) and then I will begin to advertise the project (no idea how to do that).

**Comment by mrjbq7** [comment]
Very cool!!!

**Comment by erg** [comment]
I downloaded the version from 2016-4-15 on your site to give it a try.

1) I downloaded, opened to `~/Skov`, and double-clicked--it won't launch. I got it to work by running from the command line. `./Skov.app/Contents/MacOS/skov`

2) It's not obvious what the dots on the left side do from the website. You don't have to label them, but maybe a mouseover screenshot would help make more sense?

3) When you edit labels, you should probably be able to hit escape to cancel the edit and restore the previous text. Also, pressing esc on a node that has never had text could just remove it?

4) I find myself trying to drag the unconnected nodes up and down. Seems like you could accommodate this by swapping the node order in the underlying vector. Or maybe I just like to fidget too much.

5) When you grab a node endpoint to drag, maybe you could highlight all the other node endpoints that it could attach to.

6) Saw these errors in the console, the first one many times, the second only once.

```
Type :help for debugging help.
Generic word >base does not define a method for the POSTPONE: f class.
Dispatching on object: f

Type :help for debugging help.
Sequence index out of bounds
index 0
seq   f
```

7) Is there a "take me to the text file for the source" button/link? I know you want to emphasize the nodes and graphical nature, but consider it for someone wanting to understand the complete system.

8) Somehow I managed to define a word `print` in `scratchpad` but I can't get back to scratchpad to delete it.

9) I can't stop pressing `d` to delete a node (it makes a destructor) rather than pressing `r`.

10) When you mouseover a node, instead of just showing `(remove edit help)` it should also show what kind of node it is. The shapes are not enough for me.

11) What other types of data can you add besides just Text? Support more kinds of data nodes like hashtables, vectors, etc.

Maybe there's a newer version than april in github that I can try this weekend. Looking awesome and actually feels like you could write some real code in it!

**Comment by nicolas-p** [comment]
Thank you very much Doug for this detailed feedback. I’ll answer each point.

> I downloaded the version from 2016-4-15 on your site to give it a try.
> 1) I downloaded, opened to ~/Skov, and double-clicked--it won't launch. I got it to work by running from the command line. ./Skov.app/Contents/MacOS/skov
> 
> I got an email from someone else this morning saying that he wanted to try the program but it didn’t work. I told him that it might be because of the security settings of macOS, but when I downloaded version 2016-04-15 myself it didn’t work either (I recently upgraded to macOS Sierra, it was working before). So I uploaded a new binary on GitHub a few hours ago and this one should work.
> https://github.com/nicolas-p/skov/releases/tag/2016-11-11 https://github.com/nicolas-p/skov/releases/tag/2016-11-112) It's not obvious what the dots on the left side do from the website. You don't have to label them, but maybe a mouseover screenshot would help make more sense?
> 
> I had made a mouseover screenshot but I didn’t like it so I removed it. The text on the website describes what the dots do and I was hoping people to discover by themselves. But I’ll think about it.
> 3) When you edit labels, you should probably be able to hit escape to cancel the edit and restore the previous text. Also, pressing esc on a node that has never had text could just remove it?
> 
> Good propositions, and easy to implement. Currently, if you press W by mistake, you can type a non-sensical name and Enter to make it disappear but Esc is better.
> 4) I find myself trying to drag the nodes around up and down. Maybe I just like to fidget too much.
> 
> For me, it was the number one requirement: the user should have zero control over the layout of the code. Actually I think that’s the only innovative feature. There are other visual languages available but they all require you to spend half of your time doing interior decoration instead of programming.

Manual control only existed for a few weeks before I implemented the first layout algorithm. If the algorithm makes mistakes, going back to manual control is not the right solution. The right solution is to improve the algorithm.

But I now understand why every other (nodes-and-wires based) visual language has only manual layout: automatic layout is REALLY hard. I did it because I didn’t know that.

> 5) When you grab a node to drag, maybe you could highlight all the other nodes that it could attach to.
> 
> This would be useful if inputs and outputs were typed. The system wouldn’t let you connect an object of type gadget where a number is expected. I want to implement this in the future but currently the system will just let you connect any output to any input and the problem will only be detected at runtime.
> 6) Saw these errors in the console, the first one many times, the second only once.
> 
> Type :help for debugging help.
> Generic word >base does not define a method for the POSTPONE: f class.
> Dispatching on object: f
> 
> Type :help for debugging help.
> Sequence index out of bounds
> index 0
> seq   f
> Did you identify what action you did before getting these errors? The system has been pretty error-free lately on my computer so I’m surprised.
> 7) Is there a "take me to the text file for the source" button/link? I know you want to emphasize the nodes and graphical nature, but consider it for someone wanting to understand the complete system.
> 
> As you guessed I’m not in favor of this one. I removed all the Factor source files and I don’t really consider the text files generated by Skov as « source », it's just an import/export feature. The real source is inside the image file. What is exported to the text files is not the real form of the code (but it’s close).

The best way to understand how the code is stored « inside » is to add a keyboard shortcut to environment-gadget.factor so that when you put your cursor over a node and press a key, the inspector pops up displaying the contents of this node object. I think I already have the code somewhere I you want.

> 8) Somehow I managed to define a word print in scratchpad but I can't get back to scratchpad to delete it.
> 
> There is something that I haven’t explained anywhere (because I’m not very proud of it). The interface makes it look like there is a top-level vocabulary called « ● » and that you can put sub-vocabularies in it. Well, we can say that on the Skov side of the system, things are really organized this way, but on the Factor side, « ● » is actually « scratchpad » and its sub-vocabularies are actually top-level vocabularies.

So you should see your « print » word if you click on « ● ».

> 9) I can't stop pressing d to delete a node (it makes a destructor) rather than pressing r.
> 
> Yes, this was actually the original keyboard shortcut before I added destructors. There are several conflicts like this but D is the worst:
> Delete
> Disconnect
> Destructor

The definitive solution is Apple’s new TouchBar.
(I’m starting to dream about the possibilities!)

> 10) When you mouseover a node, instead of just showing (remove edit help) it should also show what kind of node it is. The shapes are not enough for me.
> 
> Good idea, I’ll do it.
> 11) What other types of data can you add besides just Text? Support more kinds of data nodes like hashtables, vectors, etc.
> 
> Ah, this is a question I spent a long time thinking about. Skov doesn’t have too much syntax. Almost everything is a word but I found it useful to introduce a few exceptions (visual syntax):
> - text because I wanted to avoid quotation marks (which are syntax too)
> - accessors, mutators, constructors, destructors because I wanted to avoid >> << <> >< (which are equivalent to syntax in a sense, despite being part of the word's name)

Originally, I wanted special visual syntax for sequences, so you could easily create a sequence containing 10, 7, 9, 2. But sequences can be nested and there I have a problem.

The best way to formulate the problem, I think, is to say that Lisp represents code as a nested tree while Skov represents code as a flat tree. This is true for code, and I would like it to be true for data as well.

So my current opinion (subject to change) is that there should be no special syntax for sequences and hashtables, and that to create an array of numbers you would use an array constructor and a separate node for each number. I know it will potentially take a lot of space on the screen but I’ve given it a lot of thought and I think it’s the right thing to do.

> Maybe there's a newer version than april in github that I can try this weekend. Looking awesome and actually feels like you could write some real code in it!
> 
> As I wrote above, you can try today’s version (2016-11-11). The main changes since April are Retina display support, better and faster graph layout, word completion and rich output (any UI gadget can be an output).

Nicolas

**Comment by mrjbq7** [comment]
I'm looking forward to adding TouchBar support also, hopefully it's easy to integrate!

**Comment by erg** [comment]
From above, I don't want to layout the code, I just had about five nodes that I hadn't connected yet and thought it would be fun to drag them around their unconnected area. Or you could sort them. It seems useless to have nodes unless you plan to connect them, but it's the first thing I tried, and swapping them in the `unused` area isn't that big a deal imo since they're still not really connected.

New list.

1) 2016-11-11 version won't launch from the mounted disk image if you just double-click the dmg and then the skov app, but it will run from double-clicking it from a folder `~/Skov/`.

2) When you start Skov, click on the red eye thing, and the screen shifts over a slight bit. Not really important unless it is caused by some underlying issue.

3) Bug report - nodes end up on top of each other.
Create a new vocab. Create a new word `hi`. Add a `Text` that says `hello`. Add a `print` word. Connect them. Run it a couple times. Add an `append` word. Connect the `hello` to the append node. For me, I now have an `append` node with three terminals, the left two attached to `hello`. I think what happened is the `append` node ended up on top of the `print` node.

4) Bug report - `found a bug in locals`
Run like `./Skov/Skov.app/Contents/MacOS/skov`.
Create a new vocab. Create a new word. Terminal says

```
You have found a bug in locals. Please report.
args [ ]
obj  local

Type :help for debugging help.
```

5) if you try to add a word node and you type a word that does not exist, the node just disappears.

6) Sometimes, if you hit `w` then `append` and hit enter, the last `d` will register in the word name, but also a `destructor` node is created. Seems like it might happen more if you are clicking on the right side of the programming space.

7) What's the left bar called? What's the right bar called? You might want to make names for these so they are easier to talk about :) (or I need to read your page more carefully, or maybe a labeled diagram would help)

What if you had a vector node that displayed its elements live and you could edit them with the `inspector` tool?

Is there a notion of a `main` word, or are they all main?

Have you thought about how to add text input to it? Stream IO? Is there an application you want to write with it, or a device you want to have it work on? Sorry for so many questions :D

How about editing Factor code through Skov? If there's really a 1:1 mapping, maybe you could use it as a text editor? If some of the syntax is irregular in master Factor branch, but there's a branch where I finished a new Factor parser and regularized all the syntax (except functors). I did not get it to look up words/compile, but it can parse almost all the files (functors) and rewrite them to disk. This sounds perfect for your editor--navigate the vocab roots->vocabs->files, load them up, edit and rewrite to disk, and have a background git process sync to a remote server/branch.

It would be nice to see the available vocabs that you can use. Maybe a separate window or something that shows `sequences`, `math`, etc.

**Comment by nicolas-p** [comment]
1) It works fine on my computer. There's something I don't understand.

2) Well spotted. I've been having problems with the alignment of the right bar for a long time. It's because the words can have an error button or a result button on their left, but the words are still aligned horizontally like if the buttons didn't exist. I change the position of the words to align them but this make the width of vocab-gadget wrong in some special cases.

3) Shit! This is a regression. I've done so many iterations of the graph layout algorithm, I hadn't noticed that nodes connected to the same output didn't have repulsion anymore.

4) Again, it works fine on my computer.

5) It saw that as a feature. Do you see it as a bug?

6) I've just tried it several times, typing very fast, and I couldn't reproduce the problem.

7) In the code, there're called **plus-button-pile** on the left and **vocab-gadget** on the right. But I may need better official names.

8) I'll think about it

9) I haven't really considered deployed applications, so for the moment `MAIN:` is not supported. I imagine a menu somewhere that would let you choose if a word is regular/memoized/main/generic

10) I'm not sure I understand the question. There is not particular application I want to write with it. And the device I've always had in mind is the traditional desktop or laptop computer. I'm not planning to port it to the iPad for exemple.

11) There's not a 1:1 mapping, in either direction.

**Factor>Skov**
I wrote above (early August) that I had managed to automatically convert Factor code to Skov code, but only in simple cases. In the future I want to be able to convert arbitrarily complex Factor code but this will not be easy.

**Skov>Factor**
This does not really exist, or not how you think. When the code needs to be compiled, I generate `lambda` objects (defined in `locals.type`) and the quotation inside is a sequence of `multi-def` objects and the inputs and outputs are represented by `local` objects. If you wrote it as textual code (which Skov never does), it would look like this:
`[ { 1 2 3 4 } :> #0 #0 first4 :> ( #1 #2 #3 #4 ) #1 #3 2array :> #5 #2 #4 2array :> #6 #5 . #6 . ]`
So it's hardly readable Factor code.

So using Skov as a text editor for Factor code is not possible.

12) I had implemented exactly what you say some months ago (late April, I think), but I soon found what I think is a better way. If you go to the help browser and look at any word or vocab, you will see a blue circle on the top right with maybe a big dot inside. This tells you if the vocab in question is active or inactive. You can activate or deactivate a vocab by clicking this circle. Now I anticipate your next remark: the blue circle it hardly visible, it should be white, all right!
